File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/listeners/ProblemBeforeSaveListener.java
package io.github.sanyavertolet.edukate.backend.listeners;

import io.github.sanyavertolet.edukate.backend.entities.Problem;
import io.github.sanyavertolet.edukate.backend.utils.SemVerUtils;
import org.bson.Document;
import org.springframework.data.mongodb.core.mapping.event.AbstractMongoEventListener;
import org.springframework.data.mongodb.core.mapping.event.BeforeSaveEvent;
import org.springframework.stereotype.Component;
import reactor.util.function.Tuple3;

import java.util.Objects;

@Component
public class ProblemBeforeSaveListener extends AbstractMongoEventListener<Problem> {

    @Override
    public void onBeforeSave(BeforeSaveEvent<Problem> event) {
        Problem entity = event.getSource();
        Document document = event.getDocument();
        Objects.requireNonNull(document);

        String problemId = entity.getId();

        Tuple3<Integer, Integer, Integer> tuple = SemVerUtils.parse(problemId);

        document.put(SemVerUtils.majorFieldName, tuple.getT1());
        document.put(SemVerUtils.minorFieldName, tuple.getT2());
        document.put(SemVerUtils.patchFieldName, tuple.getT3());
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/savelisteners/ProblemBeforeSaveListener.kt
package io.github.sanyavertolet.edukate.backend.savelisteners

import io.github.sanyavertolet.edukate.backend.entities.Problem
import io.github.sanyavertolet.edukate.backend.utils.SemVerUtils
import org.springframework.data.mongodb.core.mapping.event.AbstractMongoEventListener
import org.springframework.data.mongodb.core.mapping.event.BeforeSaveEvent
import org.springframework.stereotype.Component

@Component
class ProblemBeforeSaveListener : AbstractMongoEventListener<Problem>() {
    override fun onBeforeSave(event: BeforeSaveEvent<Problem>) {
        val entity = event.source
        val document = requireNotNull(event.document)

        val problemId = entity.id
        val (major, minor, patch) = SemVerUtils.parse(problemId)

        document[SemVerUtils.MAJOR_FIELD_NAME] = major
        document[SemVerUtils.MINOR_FIELD_NAME] = minor
        document[SemVerUtils.PATCH_FIELD_NAME] = patch
    }
}
